File: iosApp/iosApp/Net/ConnectionHolder.swift
import Foundation
import Client

/// Process-scoped singleton holding the live `TermtasticClient` and
/// `WindowSocket`. Mirrors the Android `ConnectionHolder` singleton.
@Observable
final class ConnectionHolder {
    static let shared = ConnectionHolder()

    private(set) var client: Client.TermtasticClient?
    private(set) var windowSocket: Client.WindowSocket?

    private init() {
        client = nil
        windowSocket = nil
    }

    /// Whether the server is showing an approval dialog for this connection.
    private(set) var pendingApproval: Bool = false

    /// Set to `true` by `connect(...)` when the most recent attempt failed
    /// specifically because the server's leaf cert no longer matches the
    /// stored pin (verify-mode mismatch in the shared
    /// `PinnedHttpClientFactory`). Callers branch on this in their `catch`
    /// to show a cert-changed dialog instead of a generic error — Darwin's
    /// `URLSessionDelegate` cancels the challenge with no marker NSError,
    /// so the bridged Swift error is otherwise indistinguishable from a
    /// wrong-port / unreachable-host failure.
    private(set) var lastPinMismatch: Bool = false

    private var approvalObserver: Client.FlowObserver?

    /// Tear down any existing client and create a fresh one for the given URL.
    ///
    /// `pinnedFingerprintHex` selects between the TOFU capture mode (nil,
    /// first-connect to a new host — the leaf cert's SHA-256 is observed
    /// during the handshake and exposed via `client.observedFingerprint`)
    /// and verify mode (non-nil — `URLSessionDelegate` constant-time compares
    /// the leaf against the pin and cancels the challenge on mismatch).
    /// A mismatch additionally sets `lastPinMismatch = true` so the caller's
    /// `catch` can show the cert-changed UX instead of a generic error.
    @MainActor
    func connect(
        serverUrl: Client.ServerUrl,
        authToken: String,
        pinnedFingerprintHex: String? = nil
    ) async throws {
        disconnect()
        pendingApproval = false
        lastPinMismatch = false
        let identity = Client.ClientIdentity(
            type: "Computer",
            hostname: ProcessInfo.processInfo.hostName,
            selfReportedIp: Self.firstNonLoopbackIPv4()
        )
        let fresh = Client.TermtasticClientKt.createTermtasticClient(
            serverUrl: serverUrl,
            authToken: authToken,
            identity: identity,
            pinnedFingerprintHex: pinnedFingerprintHex
        )
        let socket = fresh.openWindowSocket()
        // Watch for PendingApproval so we can update the UI while
        // waiting for the server-side dialog to be answered.
        let observer = Client.FlowObserver()
        approvalObserver = observer
        observer.observe(flow: fresh.windowState.pendingApproval) { [weak self] value in
            DispatchQueue.main.async {
                self?.pendingApproval = (value as? Bool) == true
            }
        }
        do {
            // Phase 1: WebSocket handshake — 15 s is plenty.
            try await withTimeout(seconds: 15) {
                try await socket.awaitSessionReady()
            }
            // Phase 2: wait for the first Config envelope.  When approval
            // is pending this can take minutes, so use a generous timeout.
            try await withTimeout(seconds: 300) {
                try await socket.awaitInitialConfig()
            }
            pendingApproval = false
        } catch {
            pendingApproval = false
            observer.clear()
            approvalObserver = nil
            // Sample the side-channel set by the shared
            // PinnedHttpClientFactory.ios before we tear the client down —
            // the StateFlow value stays readable after close(), but reading
            // it here keeps the lifetime obvious.
            lastPinMismatch = (fresh.observedMismatch.value as? String) != nil
            let nsError = error as NSError
            NSLog(
                "[ConnectionHolder] connect failed: domain=%@ code=%ld desc=%@ userInfo=%@",
                nsError.domain, nsError.code,
                nsError.localizedDescription, String(describing: nsError.userInfo)
            )
            try? await socket.close()
            fresh.close()
            throw error
        }
        observer.clear()
        approvalObserver = nil
        self.client = fresh
        self.windowSocket = socket
    }

    @MainActor
    func disconnect() {
        let ws = windowSocket
        windowSocket = nil
        client?.close()
        client = nil
        Task { try? await ws?.close() }
    }

    /// Best-effort first non-loopback IPv4 address.
    private static func firstNonLoopbackIPv4() -> String? {
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let first = ifaddr else { return nil }
        defer { freeifaddrs(first) }

        var cursor: UnsafeMutablePointer<ifaddrs>? = first
        while let ifa = cursor {
            let sa = ifa.pointee.ifa_addr
            if sa?.pointee.sa_family == UInt8(AF_INET) {
                let name = String(cString: ifa.pointee.ifa_name)
                if name != "lo0" {
                    var addr = [CChar](repeating: 0, count: Int(NI_MAXHOST))
                    if getnameinfo(sa, socklen_t(sa!.pointee.sa_len),
                                   &addr, socklen_t(addr.count),
                                   nil, 0, NI_NUMERICHOST) == 0 {
                        return String(cString: addr)
                    }
                }
            }
            cursor = ifa.pointee.ifa_next
        }
        return nil
    }
}

/// Swift-friendly timeout wrapper for async operations.
private func withTimeout<T>(seconds: TimeInterval, operation: @escaping () async throws -> T) async throws -> T {
    try await withThrowingTaskGroup(of: T.self) { group in
        group.addTask { try await operation() }
        group.addTask {
            try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
            throw NSError(domain: "ConnectionHolder", code: -1,
                          userInfo: [NSLocalizedDescriptionKey: "Connection timed out"])
        }
        let result = try await group.next()!
        group.cancelAll()
        return result
    }
}

File: androidApp/src/main/kotlin/se/soderbjorn/lunamux/android/net/ConnectionHolder.kt
package se.soderbjorn.lunamux.android.net

import android.util.Log
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.withTimeout
import se.soderbjorn.lunamux.android.BuildConfig
import se.soderbjorn.lunamux.client.ClientIdentity
import se.soderbjorn.lunamux.client.ServerUrl
import se.soderbjorn.lunamux.client.LunamuxClient
import se.soderbjorn.lunamux.client.WindowSocket
import java.net.Inet4Address
import java.net.InetAddress
import java.net.NetworkInterface

/**
 * Process-scoped holder for the live [LunamuxClient]. The app is tiny
 * enough that a plain singleton is cleaner than pulling in Hilt or Koin.
 * [LunamuxApp] rebuilds this whenever the user commits a new host/port
 * from the Connect screen.
 */
object ConnectionHolder {
    @Volatile
    private var currentClient: LunamuxClient? = null

    @Volatile
    private var currentWindowSocket: WindowSocket? = null

    /**
     * Returns the currently connected [LunamuxClient], or null if disconnected.
     *
     * @return the active client instance, or null.
     */
    fun client(): LunamuxClient? = currentClient

    /**
     * Returns the open [WindowSocket] for the current connection, or null if disconnected.
     *
     * @return the active window socket, or null.
     */
    fun windowSocket(): WindowSocket? = currentWindowSocket

    /**
     * Reflects whether the server is currently showing a device-approval
     * dialog for this connection. UI can observe this to show
     * "Waiting for approval…" instead of a bare spinner.
     */
    val pendingApproval: StateFlow<Boolean>?
        get() = currentClient?.windowState?.pendingApproval

    /**
     * Tears down any existing client and creates a fresh one for [serverUrl].
     *
     * Called from [se.soderbjorn.lunamux.android.ui.HostsScreen] when the
     * user taps a host to connect. Performs a two-phase handshake: first the
     * WebSocket session (15 s timeout), then waits for the server's initial
     * Config envelope (up to 5 min to allow for device-approval dialogs).
     *
     * @param serverUrl the server URL containing host and port.
     * @param authToken the authentication token for this client.
     * @param pinnedFingerprintHex lowercase hex SHA-256 of the server's leaf
     *   cert (verify mode), or `null` to run TOFU capture on first connect.
     *   The caller can read `client().observedFingerprint.value` afterwards
     *   to persist the captured pin.
     * @return the connected [LunamuxClient] instance.
     * @throws Throwable if the connection or handshake fails. A
     *   `javax.net.ssl.SSLHandshakeException` whose cause chain contains a
     *   `CertificateException` starting with `"pin-mismatch:"` indicates the
     *   server's cert no longer matches the stored pin; UI should surface
     *   the cert-changed dialog (see [isPinMismatch]).
     */
    suspend fun connect(
        serverUrl: ServerUrl,
        authToken: String,
        pinnedFingerprintHex: String? = null,
    ): LunamuxClient {
        disconnect()
        Log.i("ConnectionHolder", "connect: ${serverUrl.wsUrl("/window")}")
        val fresh = LunamuxClient(
            serverUrl = serverUrl,
            authToken = authToken,
            // Android devices report type "Android" so the settings UI can tell
            // them apart from iOS and browser tabs, and the running app version
            // so the server can gate newer pane kinds (agent consoles, 1.5+) to
            // clients that can render them. We do a best-effort hostname + first
            // non-loopback IPv4 lookup here; both are advisory so a failure just
            // means blanks.
            identity = ClientIdentity(
                type = "Android",
                hostname = runCatching { InetAddress.getLocalHost().hostName }.getOrNull(),
                selfReportedIp = runCatching { firstNonLoopbackIpv4() }.getOrNull(),
                version = BuildConfig.VERSION_NAME,
            ),
            pinnedFingerprintHex = pinnedFingerprintHex,
        )
        val socket = fresh.openWindowSocket()
        try {
            // Phase 1: WebSocket handshake — 15 s is plenty.
            withTimeout(15_000) {
                socket.awaitSessionReady()
            }
            // Phase 2: wait for the first Config envelope. When device
            // approval is pending, the server-side dialog can take minutes
            // to be answered, so use a generous timeout. The client
            // receives a PendingApproval envelope immediately (observable
            // via [pendingApproval]) so the UI can show feedback.
            withTimeout(300_000) {
                socket.awaitInitialConfig()
            }
        } catch (t: Throwable) {
            Log.w("ConnectionHolder", "connect failed", t)
            runCatching { socket.close() }
            runCatching { fresh.close() }
            throw t
        }
        currentClient = fresh
        currentWindowSocket = socket
        return fresh
    }

    /**
     * Scans all network interfaces for the first non-loopback, non-link-local
     * IPv4 address to use as the advisory self-reported IP in the client identity.
     *
     * @return the IP address string, or null if none found.
     */
    private fun firstNonLoopbackIpv4(): String? {
        val nics = NetworkInterface.getNetworkInterfaces() ?: return null
        for (nic in nics) {
            if (!nic.isUp || nic.isLoopback) continue
            for (addr in nic.inetAddresses) {
                if (addr is Inet4Address && !addr.isLoopbackAddress && !addr.isLinkLocalAddress) {
                    return addr.hostAddress
                }
            }
        }
        return null
    }

    /**
     * App-resume hook: if the `/window` socket has been quiet long enough
     * to be presumed dead (the OS silently kills TCP connections while the
     * phone sleeps; the read loop never errors), force an immediate
     * reconnect so the tree/screens refresh with the server's current
     * state. No-op when disconnected or when the connection is healthy —
     * the server pushes session states every ~3 s, so a healthy socket is
     * never quiet for [WINDOW_STALE_MS].
     *
     * Called from [se.soderbjorn.lunamux.android.ui.LunamuxApp]'s
     * lifecycle observer on every ON_START.
     */
    fun refreshAfterResume() {
        currentWindowSocket?.reconnectIfStale(WINDOW_STALE_MS)
    }

    /** Quiet threshold for [refreshAfterResume]; see the 3 s state poller. */
    private const val WINDOW_STALE_MS = 8_000L

    /**
     * Closes the current window socket and client, resetting both to null.
     * Safe to call even when already disconnected.
     */
    suspend fun disconnect() {
        currentWindowSocket?.close()
        currentWindowSocket = null
        currentClient?.close()
        currentClient = null
    }

    /**
     * Walk a throwable's cause chain looking for the `"pin-mismatch:"` marker
     * thrown by [se.soderbjorn.lunamux.client.createPinnedHttpClient] when
     * the server's leaf cert no longer matches the stored pin. Used by
     * [se.soderbjorn.lunamux.android.ui.HostsScreen] to decide between
     * showing a generic error snackbar and a cert-changed re-pair dialog.
     *
     * @param t the throwable to inspect (typically the failure surfaced from
     *   [connect]).
     * @return `true` if any link in the cause chain is a pin-mismatch.
     */
    fun isPinMismatch(t: Throwable): Boolean {
        var c: Throwable? = t
        var depth = 0
        while (c != null && depth < 16) {
            val msg = c.message
            if (msg != null && msg.startsWith("pin-mismatch:")) return true
            c = c.cause
            depth++
        }
        return false
    }
}
